packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts

```diff
@@ -0,0 +1,176 @@
+  const [error, setError] = useState<string>();
+
+  const ethWallet = useEthWallet();
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    function getPresentationRequest() {
+      const { searchParams } = new URL(window.location.href.replace("#", ""));
+      const request = searchParams.get("request");
+
+      if (!request) {
+        return;
+      }
+
+      dispatch(fetchVerifiableCredentials());
+
+      setVerifiablePresentationRequest(request);
+    }
+    getPresentationRequest();
+  }, [setVerifiablePresentationRequest, fetchVerifiableCredentials, dispatch]);
+
+  const onCloseModal = useCallback(() => {
+    dispatch(closePopup());
+  }, [dispatch]);
+
+  const onRejectVerifiablePresentationRequest = useCallback(async () => {
+    await dispatch(rejectVerifiablePresentationRequest());
+    onCloseModal();
+  }, [rejectVerifiablePresentationRequest, dispatch, onCloseModal]);
+
+  const onToggleSelectVerifiableCredential = useCallback(
+    (selectedHash: string) => {
+      if (error) {
+        setError(undefined);
+      }
+
+      if (selectedVerifiableCredentialHashes.includes(selectedHash)) {
+        setSelectedVerifiableCredentialHashes((hashes) => hashes.filter((hash) => hash !== selectedHash));
+      } else {
+        setSelectedVerifiableCredentialHashes((hashes) => [...hashes, selectedHash]);
+      }
+    },
+    [selectedVerifiableCredentialHashes, setSelectedVerifiableCredentialHashes, error, setError],
+  );
+
+  const onConfirmVerifiableCredentialSelection = useCallback(() => {
+    if (selectedVerifiableCredentialHashes.length === 0) {
+      setError("Please select at least one credential.");
+      return;
+    }
+
+    const verifiableCredentials = cryptkeeperVerifiableCredentials
+      .filter((cryptkeeperVerifiableCredential) =>
+        selectedVerifiableCredentialHashes.includes(cryptkeeperVerifiableCredential.metadata.hash),
+      )
+      .map((cryptkeeperVerifiableCredential) => cryptkeeperVerifiableCredential.verifiableCredential);
+    const newVerifiablePresentation = generateVerifiablePresentationFromVerifiableCredentials(verifiableCredentials);
+
+    setVerifiablePresentation(newVerifiablePresentation);
+  }, [cryptkeeperVerifiableCredentials, selectedVerifiableCredentialHashes, setVerifiablePresentation, setError]);
+
+  const onReturnToVerifiableCredentialSelection = useCallback(() => {
+    setVerifiablePresentation(undefined);
+  }, [setVerifiablePresentation]);
+
+  const onConnectWallet = useCallback(async () => {
+    await ethWallet.onConnect().catch(() => {
+      setError("Wallet connection error");
+    });
+  }, [setError, ethWallet.onConnect]);
+
+  const onSubmitVerifiablePresentation = useCallback(
+    async (needsSignature: boolean) => {
+      if (!verifiablePresentation) {
+        setError("Failed to generate Verifiable Presentation.");
+        return;
+      }
+
+      if (!needsSignature) {
```

Comment:
Why cryptkeeper signature is not supported?

Reply:
As we discussed in the meeting last week, it doesn't make sense for Verifiable Credentials to be signed with a Cryptkeeper signature. These signatures must be verified by a verifier with access to a public registry of keys, and so Ethereum keys makes sense for this purpose. Signing with Cryptkeeper can be implemented if Cryptkeeper keys are included in some public registry

Reply:
Cryptkeeper uses the same standard for wallets and it has the same key pair. If I as a user tries to sign VC with metamask keys, how someone can verify the signature? User doesn't provide public keys in this flow, they just sign the VC.

Reply:
In practice, VCs would be issued to a particular Ethereum address, which would be used as an identifier for the user. Then the signature could just be checked against the address in the VC. Yes, Cryptkeeper keys may be a valid format but I don't expect issuers **and** verifiers to be aware of Cryptkeeper public keys, unless we build a registry.

Reply:
First of all, it needs to support the same user flows. For signing identity user can use both options but for VC only external wallet option. We can understand it internally but how to explain this to users?
Regarding registry: what is the registry for external wallet keys and why do we need to worry about hosting our own registry?

Reply:
Makes sense. Registry is for verifiers to know about public keys. We don't need to make our own, but at least the public keys need to be accessible to a verifier.

Reply:
I created an issue for supporting exporting public keys for cryptkeeper accounts #868